In [1]:
import pandas as pd

df = pd.read_csv("/Users/saiyambardia/heart-disease-prediction/data/cleveland_cleaned.csv")

In [2]:
df.head()

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,target
0,63.0,1.0,1.0,145.0,233.0,1.0,2.0,150.0,0.0,2.3,3.0,0.0,6.0,0
1,67.0,1.0,4.0,160.0,286.0,0.0,2.0,108.0,1.0,1.5,2.0,3.0,3.0,1
2,67.0,1.0,4.0,120.0,229.0,0.0,2.0,129.0,1.0,2.6,2.0,2.0,7.0,1
3,37.0,1.0,3.0,130.0,250.0,0.0,0.0,187.0,0.0,3.5,3.0,0.0,3.0,0
4,41.0,0.0,2.0,130.0,204.0,0.0,2.0,172.0,0.0,1.4,1.0,0.0,3.0,0


In [3]:
df.shape

(303, 14)

In [4]:
X = df.drop(['target'], axis=1)
y = df.target

In [5]:
from sklearn.model_selection import train_test_split

X_train, X_valid, y_train, y_valid = train_test_split(X,y,test_size=0.2, train_size=0.8, random_state=0)

print(f"Shape of X_train {X_train.shape}")
print(f"Shape of X_valid {X_valid.shape}")
print(f"Shape of y_train {y_train.shape}")
print(f"Shape of y_valid {y_valid.shape}")

Shape of X_train (242, 13)
Shape of X_valid (61, 13)
Shape of y_train (242,)
Shape of y_valid (61,)


## 1. Logistic Regression

In [9]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler

pipeline = Pipeline(steps=[
    ('scaler', StandardScaler()),
    ('model', LogisticRegression(max_iter=1000))
])

pipeline.fit(X_train,y_train)
y_pred = pipeline.predict(X_valid)

In [10]:
print(f"Accuracy Score: {accuracy_score(y_valid, y_pred):.2f}\n")
print("Confusion Matrix:")
print(confusion_matrix(y_valid, y_pred))
print("\nClassification Report:")
print(classification_report(y_valid, y_pred))

Accuracy Score: 0.79

Confusion Matrix:
[[31  4]
 [ 9 17]]

Classification Report:
              precision    recall  f1-score   support

           0       0.78      0.89      0.83        35
           1       0.81      0.65      0.72        26

    accuracy                           0.79        61
   macro avg       0.79      0.77      0.78        61
weighted avg       0.79      0.79      0.78        61



Observations:
- The `LogisticRegression` model gives us an accuracy score of 0.79.
- The confusion matrix shows us that model falsely identifies 9 patients with CAD as healthy and 4 patients with no CAD are identified as with CAD
- Model is better at finding healthy patients (recall 0.89) than CAD patients (recall 0.65)
- CAD recall of 0.65 means the model misses 35% of actual CAD cases